### PR TITLE
[TP] fully rewrite Tensor Parallel APIs

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114732

This PR rewrites Tensor Parallel implementation. Tensor Parallel APIs
supposed to be a very thin-wrapper to DTensor APIs, but the current
implementation got too messy and buggy. It's really hard to debug what
went wrong when using it. It's crucially important for advanced users or
developers to understand the API and its implementation easily without
going through all different types of functions and utils, so that
they could trust what happen under the hood.

In particular this PR:

* Make ParallelStyle to be a real contract API for parallelize_module to
  take, each concrete ParallelStyle only needs to implement `apply` to
apply the sharding to nn.Module, remove all non-necessary fields. This
also enable easier ParallelStyle authoring going forward.
* Keep the ColwiseParallel and RowwiseParallel public interface, but
  refactor them in a way that makes the parameter sharding, inputs and
outputs handling lives within the style itself, so that it's easy to
understand how Linear/Embedding layers are sharded and how the inputs/outputs
transformations are performed.
* remove all those private _prepare_input/_prepare_output_fn fields for
  both ColwiseParallel/RowwiseParallel. Since we throw deprecation
messages in nightly for a while and TP is on prototype release, the
fields are also private, it should be safe to remove them
* Refactor the recently landed PrepareModuleInput/Output style, change
  output_layouts to desired_input/output_layouts, group
  the function inside the style itself, no default arguments for these
two styles and user need to specify them to think about the sharding
layouts. Fixed bugs about not handling
`use_local_output` flag.
* Make default arguments be None instead of Placement object, this is
  standard python practice to not have custom object instance as default
argument
* Remove all dead APIs (i.e. PairwiseParallel and SequenceParallel
  style, all prepare input/output functions) as we throw deprecation
 msgs for a while, and in the progress of removing all of them from the tests.
* throw deprecation warning for `tp_mesh_dim` as we recomemnd use device
  mesh slice/indexing instead of manually specify mesh dim
* Rewrite all documentations for every ParallelStyle and make the
  documentation more clear about what each style is doing

TODOs:
* Rewrite TP tests to adjust for the changes we have in this PR
* add more tests to guard the bug fixes

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

Differential Revision: [D51761183](https://our.internmc.facebook.com/intern/diff/D51761183)

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -1,652 +1,357 @@
...
-            _prepare_output=prepare_output_fn,
-        )
-
-    @staticmethod
-    def _prepare_input_fn(input_layouts, output_layouts, inputs, device_mesh=None):
-        return (
-            _redistribute_per_both_layouts(
-                inputs[0], input_layouts, output_layouts, device_mesh
-            ),
+    def _prepare_output_fn(self, outputs, device_mesh):
```
Comment (wz337) [inline torch/distributed/tensor/parallel/style.py:115]: Super weird that the Python highlight is not rendered correctly here, but when I clicked on view file, everything seems ok.

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -1,652 +1,357 @@
...
-
-    When the input is not a :class:`torch.Tensor` or :class:`DTensor`, if no layout is
-    specified, it will be a no-op. Otherwise, it will throw an error.
+    Annotate the input tensors of nn.Module to DTensors according to input_layouts, and perform layout redistribution
+    according to the desired_input_layouts.
+
+    Keyword Args:
+        input_layouts (Union[Placement, Tuple[Placement]]):
+            The DTensor layouts of input tensors for the nn.Module, this is used to annotate the input tensors to
+            become a DTensor. if some inputs are not tensor, `None` need to be specified as a placeholder.
```
Comment (wz337) [inline torch/distributed/tensor/parallel/style.py:227]: Nit: if -> If

File: torch/distributed/tensor/parallel/style.py [hunk 3]
```diff
@@ -1,652 +1,357 @@
...
-
-    When the output is not a :class:`DTensor`, if no layout is specified, it will be
-    a no-op. Otherwise, it will throw an error.
+    Annotate the output tensors of nn.Module with DTensors according to output_layouts, and perform layout redistribution
+    according to the desired_output_layouts.
+
+    Keyword Args:
+        output_layouts (Union[Placement, Tuple[Placement]]):
+            The DTensor layouts of output tensors for the nn.Module, this is used to annotate the output tensors to
+            a DTensor if they are `torch.Tensor`. if some outputs are not tensor, `None` need to be specified as a placeholder.
```
Comment (wz337) [inline torch/distributed/tensor/parallel/style.py:295]: Nit: if -> If

**Review by wz337 — APPROVED** [review]
Overall LGTM! Should we wait for TIanyu's PR for deprecating PairwiseParallel to land first to avoid any ci breakage?

**Comment by wz337** [comment]
Also, we want to remove the PairwiseParallel import from here: https://github.com/pytorch/pytorch/blob/d71de9770e96ec0f34b34c1f7a3e5dd7a82a55d1/torch/distributed/_tensor/examples/checkpoint_example.py#L25, if it is not already captured by Tianyu's PR.

File: torch/distributed/tensor/parallel/style.py [hunk 4]
```diff
@@ -1,652 +1,357 @@
...
-
-    Args:
-        input_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which DTensor will be created upon.
-        output_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which created DTensor will be redistributed to.
-        use_local_output (bool):
-            Whether to convert the DTensor to local :class:`torch.Tensor`.
-
+    Partition the column of a nn.Module and allow the parallelized module to run colwise sharded computation.
```
Comment (wconstab) [inline torch/distributed/tensor/parallel/style.py:33]: nit: the phrasing doesn't make sense to me?  ("column of a nn.module") 

I guess you mean "partition an nn.Module in a column-wise fashion"?  (that's more clear but still could use more defining what it means or if it assumes the nn.Module is of a certain shape/format.  (must be a single Linear?)

File: torch/distributed/tensor/parallel/style.py [hunk 5]
```diff
@@ -1,652 +1,357 @@
...
-
-    Args:
-        input_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which DTensor will be created upon.
-        output_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which created DTensor will be redistributed to.
-        use_local_output (bool):
-            Whether to convert the DTensor to local :class:`torch.Tensor`.
-
+    Partition the row of a nn.Module and allow the parallelized module to run rowwise sharded computation.
```
Comment (wconstab) [inline torch/distributed/tensor/parallel/style.py:134]: similar to my comment for Colwise, and after i read more of the docs/code, i think something like this makes sense

Partition a compatible nn.Module in a row-wise fashion.  Currently supports nn.Linear only.

File: torch/distributed/tensor/parallel/style.py [hunk 6]
```diff
@@ -1,652 +1,357 @@
...
-
-    :class:`PrepareModuleInput` enables users to annotate :class:`torch.Tensor` or :class:`DTensor`
-    inputs with ``input_layouts`` and ``output_layouts`` so that each input can be converted to
-    :class:`DTensor` based on the annotation. Specifically, a DTensor will be created
-    from the input Tensor based on ``input_layouts`` and then redistributed to another
-    DTensor based on ``output_layouts``.
-
-    When the input is not a :class:`torch.Tensor` or :class:`DTensor`, if no layout is
-    specified, it will be a no-op. Otherwise, it will throw an error.
+    Annotate the input tensors of nn.Module to DTensors according to input_layouts, and perform layout redistribution
```
Comment (wconstab) [inline torch/distributed/tensor/parallel/style.py:221]: the verb here isn't quite clear.  Annotate sounds like it just marks something.  But "to DTensors" sounds like it made a change (conversion).  Would it be more correct to say "Convert the input tensors of nn.Module to DTensors"?

or maybe this is more correct? 

"Configure the nn.Module's to convert its inputs to DTensor at runtime following the parallelize plan"

**Review by wconstab — COMMENTED** [review]
looks like a good PR to me.  I didn't go too deep but I am happy to see the reduction in APIs.

File: torch/distributed/tensor/parallel/style.py [hunk 7]
```diff
@@ -1,652 +1,357 @@
...
-        output_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which created DTensor will be redistributed to.
-        use_local_output (bool):
-            Whether to convert the DTensor to local :class:`torch.Tensor`.
-
+    Partition the column of a nn.Module and allow the parallelized module to run colwise sharded computation.
+    Users can compose it with RowwiseParallel to achieve the sharding of more complicated modules.
+
+    Keyword Args:
+        input_layouts (Placement, optional):
```
Comment (awgu) [inline torch/distributed/tensor/parallel/style.py:37]: Would it be possible to call this arg `input_placement` (singular) (and similarly `output_placement`) since the type is of `Placement`? Is the concern backward compatibility?

(Or, if we want consistency with `PrepareModuleInput`, whose `input_layouts` accepts `Tuple[Placement, ...]`, then we can name them all `input_placements` (plural)?)

Comment (wanchaol) [reply torch/distributed/tensor/parallel/style.py:37]: `input_placements` is sth I've considered, the reason I ended up not using them is that I would like to avoid user confusion when they directly using DTensor APIs, `placements` is a specific term to DTensor where it's a List/Tuple of placement that "corresponding to Device Mesh dimensions". However here for `PrepareModuleInput/Output`, it also have a Tuple of Placement, but it corresponds to "multiple tensors/inputs of the  nn.Module", so user might be confused if the naming are the same.. lmk if this make sense

Comment (awgu) [reply torch/distributed/tensor/parallel/style.py:37]: Ohh, I see. That makes sense! Thanks for the explanation!

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51761183).

File: torch/distributed/tensor/parallel/style.py [hunk 8]
```diff
@@ -1,652 +1,369 @@
...
-            prepare_input_fn = _prepare_input
-        elif input_layouts is not None:
-            prepare_input_fn = functools.partial(
-                RowwiseParallel._prepare_input_fn,
-                input_layouts,
-                Shard(-1),
-            )
+        super().__init__()
+        self.input_layouts = (input_layouts or Replicate(), )
+        self.output_layouts = (output_layouts or Shard(-1), )
```
Comment (fduwjj) [inline torch/distributed/tensor/parallel/style.py:74]: Do we still want to keep the check of type of input_layouts and output_layouts?

**Review by fduwjj — APPROVED** [review]
Thanks for cleaning up and make TP usability better. I think overall this looks good to me and please make sure all test passed.
